pkg/acpi/coreboot/madt.go

```diff
@@ -0,0 +1,169 @@
+		j := len(b) - i - 1
+		b[i], b[j] = b[j], b[i]
+	}
+	return b
+}
+
+// NewMADT returns a MADT, from a Table, iff the signature is correct.
+func NewMADT(t acpi.Table) (Corebooter, error) {
+	if t.Sig() != sigMADT {
+		return nil, fmt.Errorf("%s is not a MADT", t.Sig())
+	}
+	b := t.TableData()
+	Debug("NEWMADT Subtables: %d bytes", len(b))
+	if len(b) < int(minMADT) {
+		return nil, fmt.Errorf("Table is only %d bytes, need %d", len(t.TableData()), minMADT)
+	}
+	b = b[8:]
+	Debug("NEWMADT Subtables: after removing 8 bytes of fluff, %d bytes", len(b))
+	m := &MADT{Table: t, base: t.Address()}
+	for len(b) > 0 {
+		Debug("NewMADT loop: %d bytes\n", len(b))
+		if len(b) < 2 {
+			return nil, fmt.Errorf("subtable is too short: have %d bytes, need 2", len(b))
+		}
+		l := b[1]
+		typ := madtType(b[0])
+		Debug("subtable: %d bytes, l is %#x, t is %#x", len(b), l, t)
+		if int(typ) > len(info) || typ == entryBOGUS {
+			Debug("Bad type: %#x, offset %#x, skipping %d bytes", typ, len(t.TableData())-len(b), l)
+			b = b[l:]
+			continue
+		}
+		if l != info[typ].size {
+			return nil, fmt.Errorf("Type %s has size %d should be %d", info[typ].name, l, info[typ].size)
+		}
+		if len(b) < int(l) {
+			return nil, fmt.Errorf("Type %s has buf length %d should be %d", info[typ].name, len(b), info[typ].size)
+		}
+		m.subtables = append(m.subtables, subtable{t: typ, b: b})
+		b = b[l:]
+	}
+	Debug("NEWMADT Subtables: %d bytes tab %d bytes", len(b), len(m.TableData()))
+	return m, nil
+}
+
+// CoreBoot generates the coreboot C code for a MADT.
+// MADT consists of TLV entries: 1-byte type, 1-byte length, value.
+func (m *MADT) Coreboot(w io.Writer) error {
+	s := fmt.Sprintf(`unsigned long acpi_fill_madt(unsigned long current)
+{
+	/* create all subtables for processors */
+`)
+	// OK this is a bit nasty. Right now coreboot is trying to fill out
+	// the lapic address and flags. So we ignore those in the table for now.
+	// If it's getting them wrong, we will have to go back and
+	// fill them in.
+	//lapic(%#x, %#x);\n", b[:4], b[4:8])
+	// The coreboot serialization code for ACPI uses the C compiler to drive serialization.
+	// It uses this pattern:
+	// int acpi_create_madt_lapic(acpi_madt_lapic_t *lapic, u8 cpu, u8 apic);
+	// The first arg is a uintptr cast to a struct pointer of the table type.
+	// The function stores the arguments into the packed struct, the idea being that
+	// the compiler generates code to store into the struct that exactly matches
+	// the ACPI standard.
+	for _, tab := range m.subtables {
+		b := tab.b
+		switch tab.t {
+		case entryLAPIC:
+			// TODO: Check b[4] to see if the LAPIC is enabled? When would it not be, if it's in the table?
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_lapic((acpi_madt_lapic_t *)current, %#x, %#x);\n", b[2], b[3])
+		case entryIOAPIC:
+			s += fmt.Sprintf("\tcurrent += acpi_create_madt_ioapic((acpi_madt_ioapic_t *)current, %#x, %#x, %#x);\n", b[2], rv(b[4:8]), rv(b[8:12]))
```

Comment:
regarding the rv() stuff, is that an endian thing?  getting them in host-endian, to get the right constants in the C code.  (don't know what happens when you pass a byte slice to a %#x).

Reply:
coreboot code seems to want the constants this way.

I'm not an actual fan of how coreboot did all this but ... it won't change.

This is just converting a byte slice to an order s.t. when I print it with %#x it prints right. 

I think this is too lazy for words and probably ought to be changed. But this patch may go on ice anyway, it's not proving to be useful in practice.

Reply:
no problem.  don't take my questions as blockers.  i saw the commit fly by, and since i know a little about the MADT, thought i'd use it as an opportunity to look at some go code.

Reply:
should we still keep reviewing this, ron?